driver-core/src/main/com/mongodb/client/model/expressions/Branches.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.client.model.expressions;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Function;
+
+public final class Branches {
+
+    Branches() {
+        // package-private
+    }
+
+    private <T extends Expression, R extends Expression> BranchesIntermediary<T, R> with(final Function<T, SwitchCase<R>> switchCase) {
+        List<Function<T, SwitchCase<R>>> v = new ArrayList<>();
+        v.add(switchCase);
+        return new BranchesIntermediary<>(v);
+    }
+
+    private static <T extends Expression> MqlExpression<?> mqlEx(final T value) {
+        return (MqlExpression<?>) value;
+    }
+
+    // is fn
+
+    public <T extends Expression, R extends Expression> BranchesIntermediary<T, R> is(final Function<T, BooleanExpression> o, final Function<T, R> r) {
+        return this.with(value -> new SwitchCase<>(o.apply(value), r.apply(value)));
+    }
+
+    // eq lt lte
```

Comment:
Am I correct in concluding that all of the below methods are expressible using the above `is` method?

Reply:
Yes. These, and any others that might be added to these classes, are all convenience methods.

However, the isBoolean (and other type) methods perform a cast that is not externally permitted by our API, so that the right lambda receives the value as a boolean (this is like instanceof pattern matching in Java). It would be verbose for the user otherwise.

Reply:
Got it.  I was playing around with this and not sure how to express something like this (pseudocode):

```
switch
   case isBoolean b and b is true ...
   case isBoolean b and b is false ...
   case isString s and s is null or empty ...
   case isString s ...
```

I can't quite see how it works given the type system enforced by `BranchIntermediary`. Can you clarify?

Reply:
I'm unsure how to best express this without knowing the `...` (are they more checks, or the then branch, and if so, about how many lines are ensuing) - I've assumed they are short return values. In Java, this would be:

```
if (value instanceof Boolean b) {
    return b ? "one" : "two";
} else if (value == null) {
    return "three";
} else if (value instanceof String s) {
    return "four";
}
```
So in the API:
```
expr.switchMap(on -> on
        .isBoolean(b -> b.cond(of("one"), of("two")))
        .isNull(v -> of("three"))
        .isString(s -> of("four")))
```

As with Java, `null` is not considered a String value, and the same for `missing`, so the third case in the original example is never taken. (Note that the API is non-null, and attempts to forbid even the assignment of the null value to a string.)

It seems that complex cases can always be expressed via nesting (comparable to nested if statements), and common code can be extracted into a lambda or method:

```
expr.switchMap(on -> on
    .isString(s -> s.switchMap(on2 -> on2
        .eq(v1, lambda1)
        .eq(v2, lambda2)
    .isBoolean(b -> ...
```

Since everything can be expressed, the remaining question is whether we can offer a better second way of expressing the above. When I've fleshed out possible examples, I've found they don't make sense (often because Java has an infinite number of types that can be instance-of'd, but we have < 10, so the reasonable usages are different), or the examples are at least equally-well expressed using what we have in the API.